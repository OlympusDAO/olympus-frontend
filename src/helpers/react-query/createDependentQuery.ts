import { QueryKey, useQuery } from "react-query";

import { nonNullable } from "../types/nonNullable";

/**
 * Used to build a `useQuery` function for fetching necessary data in parallel for a query,
 * using that queries `queryKey`
 *
 * Please refer to the `useStakePoolTVL` function for an example on why this function is handy.
 */
export const createDependentQuery = (baseQueryKey: QueryKey) => {
  return <TData>(key: string, fn: () => Promise<TData>, enabled?: boolean) => {
    const _key = [...baseQueryKey, key].filter(nonNullable);

    return useQuery(_key, fn, { enabled }).data;
  };
};
