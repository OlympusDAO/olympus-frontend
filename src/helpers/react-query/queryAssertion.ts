/**
 * Assertion function helpful for asserting `enabled`
 * values from within a `react-query` function.
 * @param value The value(s) to assert
 * @param queryKey Key of current query
 */
export function queryAssertion(value: unknown, queryKey: any = "not specified"): asserts value {
  if (!value) throw new Error(`Failed react-query assertion for key: ${queryKey}`);
}
