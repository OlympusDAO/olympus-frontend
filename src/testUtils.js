import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "@material-ui/core/styles";
import CssBaseline from "@material-ui/core/CssBaseline";
import { light as lightTheme } from "./themes/light.js";
import { Web3ContextProvider } from "./hooks/web3Context";
import store from "./store.ts";
import { createMemoryHistory } from "history";
import App from "./App";
import { I18nProvider } from "@lingui/react";
import { i18n } from "@lingui/core";

const AllTheProviders = ({ children }) => {
  return (
    <Web3ContextProvider>
      <Provider store={store}>
        <BrowserRouter basename={"/"}>
          <ThemeProvider theme={lightTheme}>
            <CssBaseline />
            {children}
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    </Web3ContextProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

const renderRoute = function (route) {
  const history = createMemoryHistory();
  history.push(route);
  render(
    <Web3ContextProvider>
      <Provider store={store}>
        <I18nProvider i18n={i18n}>
          <BrowserRouter basename={"/#"}>
            <App />
          </BrowserRouter>
        </I18nProvider>
      </Provider>
    </Web3ContextProvider>,
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, renderRoute };
