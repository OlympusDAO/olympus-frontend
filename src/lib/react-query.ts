import { QueryClient } from "react-query";

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnMount: false,
      refetchOnWindowFocus: false,
      notifyOnChangeProps: "tracked",
    },
  },
});
