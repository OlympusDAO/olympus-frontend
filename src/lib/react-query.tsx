import React from "react";
import { QueryCache, QueryClient, QueryClientProvider } from "react-query";
import { ReactQueryDevtools } from "react-query/devtools";
import { Environment } from "src/helpers/environment/Environment/Environment";

const queryCache = new QueryCache({
  onError: (error, query) => {
    if (error instanceof Error) console.error({ key: query.queryKey, error: error.message });
  },
});

const queryClient = new QueryClient({
  queryCache,
  defaultOptions: {
    queries: {
      refetchOnMount: false,
      refetchInterval: false,
      refetchOnReconnect: false,
      refetchOnWindowFocus: false,
      notifyOnChangeProps: "tracked",
    },
  },
});

export const ReactQueryProvider: React.FC = ({ children }) => (
  <QueryClientProvider client={queryClient}>
    {Environment.env.NODE_ENV === "development" && <ReactQueryDevtools />}

    {children}
  </QueryClientProvider>
);
