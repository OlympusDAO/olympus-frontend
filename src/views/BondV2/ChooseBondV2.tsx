import "./ChooseBond.scss";

import { t, Trans } from "@lingui/macro";
import {
  Box,
  Grid,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
  Zoom,
} from "@material-ui/core";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import { Metric, MetricCollection, Paper } from "@olympusdao/component-library";
import isEmpty from "lodash/isEmpty";
import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router";
import { useAppSelector, useWeb3Context } from "src/hooks";
import { usePathForNetwork } from "src/hooks/usePathForNetwork";
import { IUserBondDetails } from "src/slices/AccountSlice";
import { getAllBonds, getUserNotes, IUserNote } from "src/slices/BondSliceV2";
import { AppDispatch } from "src/store";

import { formatCurrency } from "../../helpers";
import { BondDataCard, BondTableData } from "./BondRow";
import ClaimBonds from "./ClaimBonds";

function ChooseBondV2() {
  const { networkId, address, provider } = useWeb3Context();
  const dispatch = useDispatch<AppDispatch>();
  const history = useHistory();
  usePathForNetwork({ pathName: "bonds", networkID: networkId, history });

  const bondsV2 = useAppSelector(state => {
    return state.bondingV2.indexes.map(index => state.bondingV2.bonds[index]).sort((a, b) => b.discount - a.discount);
  });

  const isSmallScreen = useMediaQuery("(max-width: 733px)"); // change to breakpoint query
  const accountNotes: IUserNote[] = useAppSelector(state => state.bondingV2.notes);

  const marketPrice: number | undefined = useAppSelector(state => {
    return state.app.marketPrice;
  });

  const treasuryBalance = useAppSelector(state => state.app.treasuryMarketValue);

  const isBondsLoading = useAppSelector(state => state.bondingV2.loading ?? true);

  const formattedTreasuryBalance = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    maximumFractionDigits: 0,
    minimumFractionDigits: 0,
  }).format(Number(treasuryBalance));

  useEffect(() => {
    const interval = setTimeout(() => {
      dispatch(getAllBonds({ address, networkID: networkId, provider }));
      dispatch(getUserNotes({ address, networkID: networkId, provider }));
    }, 60000);
    return () => clearTimeout(interval);
  });

  const v1AccountBonds: IUserBondDetails[] = useAppSelector(state => {
    const withInterestDue = [];
    for (const bond in state.account.bonds) {
      if (state.account.bonds[bond].interestDue > 0) {
        withInterestDue.push(state.account.bonds[bond]);
      }
    }
    return withInterestDue;
  });

  return (
    <div id="choose-bond-view">
      {(!isEmpty(accountNotes) || !isEmpty(v1AccountBonds)) && <ClaimBonds activeNotes={accountNotes} />}

      <Zoom in={true}>
        <Paper headerText={`${t`Bond`} (4,4)`}>
          <MetricCollection>
            <Metric
              label={t`Treasury Balance`}
              metric={formattedTreasuryBalance}
              isLoading={!!treasuryBalance ? false : true}
            />
            <Metric
              label={t`OHM Price`}
              metric={formatCurrency(Number(marketPrice), 2)}
              isLoading={marketPrice ? false : true}
            />
          </MetricCollection>

          {bondsV2.length == 0 && !isBondsLoading && (
            <Box display="flex" justifyContent="center" marginY="24px">
              <Typography variant="h4">No active bonds</Typography>
            </Box>
          )}

          {!isSmallScreen && bondsV2.length != 0 && (
            <Grid container item>
              <TableContainer>
                <Table aria-label="Available bonds">
                  <TableHead>
                    <TableRow>
                      <TableCell align="center">
                        <Trans>Bond</Trans>
                      </TableCell>
                      <TableCell align="left">
                        <Trans>Price</Trans>
                      </TableCell>
                      <TableCell align="left">
                        <Trans>ROI</Trans>
                      </TableCell>
                      <TableCell align="left">
                        <Trans>Duration</Trans>
                      </TableCell>
                    </TableRow>
                  </TableHead>
                  <TableBody>
                    {bondsV2.map(bond => {
                      if (bond.displayName !== "unknown")
                        return <BondTableData networkId={networkId} key={bond.index} bond={bond} />;
                    })}
                  </TableBody>
                </Table>
              </TableContainer>
            </Grid>
          )}
          <Box mt={2} className="help-text">
            <em>
              <Typography variant="body2">
                Important: New bonds are auto-staked (accrue rebase rewards) and no longer vest linearly. Simply claim
                as sOHM or gOHM at the end of the term.
              </Typography>
            </em>
          </Box>
        </Paper>
      </Zoom>

      {isSmallScreen && (
        <Box className="ohm-card-container">
          <Grid container item spacing={2}>
            {bondsV2.map(bond => {
              return (
                <Grid item xs={12} key={bond.index}>
                  <BondDataCard key={bond.index} bond={bond} networkId={networkId} />
                </Grid>
              );
            })}
          </Grid>
        </Box>
      )}
    </div>
  );
}

export default ChooseBondV2;
