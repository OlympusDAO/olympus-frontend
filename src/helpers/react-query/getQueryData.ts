import { hashQueryKey, QueryKey } from "react-query";
import { queryCache, queryClient } from "src/lib/react-query";

/**
 * A helper function that allows us to imperatively get the data
 * for a specific query whilst deduping requests for that data.
 */
export const getQueryData = async <TData>(queryKey: QueryKey, queryFn: () => Promise<TData>): Promise<TData> => {
  return await new Promise(async (resolve, reject) => {
    const state = queryClient.getQueryState<TData, Error>(queryKey);

    if (!state) {
      // If this query doesn't exist - i.e. they're are no instances
      // of `useQuery` mounted in our app yet with this specific
      // query key - we synchronously fire off a call to fetch the
      // query and cache its result against the provided queryKey.
      queryClient.prefetchQuery(queryKey, queryFn);
      // This is safe to do since `prefetchQuery` is idempotent and
      // can be called multiple times without firing off a new
      // request if one is already in-flight, allowing us to dedupe
      // multiple subscribers request for data
    }

    // Opinionated: returning stale data is better than erroring
    if (state?.status === "error") return state.data ? resolve(state.data) : reject(new Error(state.error?.message));

    // I.e. not in an error state, and data exists
    if (state?.data) return resolve(state.data);

    const hashedKey = hashQueryKey(queryKey);
    const unsubscribe = queryCache.subscribe(event => {
      if (event?.query.queryHash === hashedKey) {
        if (event.type === "queryUpdated") {
          unsubscribe();

          const { state } = event.query;

          return state.data ? resolve(state.data) : reject(new Error(state.error?.message));
        }
      }
    });
  });
};
