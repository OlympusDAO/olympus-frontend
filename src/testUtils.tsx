import { i18n } from "@lingui/core";
import { I18nProvider } from "@lingui/react";
import CssBaseline from "@material-ui/core/CssBaseline";
import { ThemeProvider } from "@material-ui/core/styles";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

import App from "./App";
import { Web3ContextProvider } from "./hooks/web3Context";
import { ReactQueryProvider } from "./lib/react-query";
import defaultStore from "./store";
import { light as lightTheme } from "./themes/light.js";

const customRender = (ui: ReactElement, store = defaultStore, options?: RenderOptions): RenderResult => {
  const ProviderWrapper = ({ children }: { children?: ReactNode }) => (
    <Web3ContextProvider>
      <ReactQueryProvider>
        <Provider store={store}>
          <I18nProvider i18n={i18n}>
            <BrowserRouter basename={"/"}>
              <ThemeProvider theme={lightTheme}>
                <CssBaseline />
                {children}
              </ThemeProvider>
            </BrowserRouter>
          </I18nProvider>
        </Provider>
      </ReactQueryProvider>
    </Web3ContextProvider>
  );
  return render(ui, { wrapper: ProviderWrapper, ...options });
};

const renderRoute = function (route: string, store = defaultStore) {
  const history = createMemoryHistory();
  history.push(route);
  return render(
    <Web3ContextProvider>
      <ReactQueryProvider>
        <Provider store={store}>
          <I18nProvider i18n={i18n}>
            <BrowserRouter basename={"/"}>
              <App />
            </BrowserRouter>
          </I18nProvider>
        </Provider>
      </ReactQueryProvider>
    </Web3ContextProvider>,
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, renderRoute };
