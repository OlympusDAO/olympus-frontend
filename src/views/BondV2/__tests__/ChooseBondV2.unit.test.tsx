import { QueryClient, QueryClientProvider } from "react-query";

import { render } from "../../../testUtils";
import ChooseBondV2 from "../ChooseBondV2";
const queryClient = new QueryClient();

describe("<ChooseBondV2/>", () => {
  it("should render component", () => {
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <ChooseBondV2 />
      </QueryClientProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
