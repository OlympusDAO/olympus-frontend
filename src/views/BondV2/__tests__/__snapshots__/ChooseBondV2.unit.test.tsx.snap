// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseBondV2/> should render component 1`] = `
<div>
  <div
    id="choose-bond-view"
  >
    <div
      class="MuiPaper-root makeStyles-root-1 makeStyles-root-2  MuiPaper-elevation0 MuiPaper-rounded"
      style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
      >
        <div
          class="MuiGrid-root card-header MuiGrid-item"
        >
          <div
            class="MuiBox-root MuiBox-root-3"
          >
            <div
              class="MuiBox-root MuiBox-root-4"
            >
              <h5
                class="MuiTypography-root header-text MuiTypography-h5"
              >
                Bond (4,4)
              </h5>
            </div>
            <div
              class="top-right"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiBox-root MuiBox-root-5"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-flex-end"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
              >
                <div
                  class="makeStyles-root-6 "
                >
                  <div
                    class="MuiBox-root MuiBox-root-7"
                  >
                    <h5
                      class="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary"
                    >
                      Treasury Balance
                    </h5>
                    <h4
                      class="MuiTypography-root MuiTypography-h4"
                      style="width: 100%;"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                        style="width: 100%;"
                      />
                    </h4>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
              >
                <div
                  class="makeStyles-root-6 "
                >
                  <div
                    class="MuiBox-root MuiBox-root-8"
                  >
                    <h5
                      class="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary"
                    >
                      OHM Price
                    </h5>
                    <h4
                      class="MuiTypography-root MuiTypography-h4"
                      style="width: 100%;"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                        style="width: 100%;"
                      />
                    </h4>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-9"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4"
            >
              No active bonds
            </h4>
          </div>
          <div
            class="MuiBox-root MuiBox-root-10 help-text"
          >
            <em>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Important: New bonds are auto-staked (accrue rebase rewards) and no longer vest linearly. Simply claim as sOHM or gOHM at the end of the term.
              </p>
            </em>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
